Project tmx level2

t-cordonnier opened the pull request:
Use TMX 1.4 level 2 for project-save.tmx

concerning team work, I implemented also the algorithm which I had suggested a few days ago:
- when you create a new project, or as long as there is not yet a project-save.tmx, the first one is saved as 1.4 Level 2 with correct tags;
- when you open an existing project, if it contains reference to DTD 1.1,
--- you load it using the old format, taking tags as they are in OmegaT < 6.1
--- when you save, you continue with the old format if you are in a team project, to avoid conflicts in Git
--- if the project is not a team one, you definitively switch to new format
- if you open a project which is already 1.4 Level 2, you continue with it
Note that I migrated also to 1.4, because this time I use the DTD version number to check whenever I am reading an old-style project (in which case I will continue with this format for better compatibility) or not.
Older versions of OmegaT will correctly read the 1.4 file, they will only continue to write it in old format but I reduce the potential impact on team working by using the just described algorithm: logically after 2-3 iterations everybody will use 1.1 in this case.




## Pull request type

feature

## Which ticket is resolved?

Discussion https://sourceforge.net/p/omegat/mailman/message/59208059/ (will be completed later)

t-cordonnier committed:
Save files as TMX level 2

t-cordonnier committed:
Compatibility with old format

t-cordonnier committed:
Keep old format only for team projects

miurahr commented on src/org/omegat/util/TMXReader2.java:
```diff
@@ -158,6 +159,10 @@ public void readTMX(File file, final Language sourceLanguage, final Language tar
+                    break;
                 case XMLEvent.START_ELEMENT:
                     StartElement eStart = (StartElement) e;
                     if ("tu".equals(eStart.getName().getLocalPart())) {
```
Can we add a callback when we get the start element of "tmx"?

t-cordonnier commented on src/org/omegat/core/team2/TeamTool.java:
```diff
@@ -106,7 +106,7 @@ public static void initTeamProject(File dir, String srcLang, String trgLang) thr
         ProjectTMX projectTMX = new ProjectTMX();
         projectTMX.load(props.getSourceLanguage(), props.getTargetLanguage(), props.isSentenceSegmentingEnabled(), null, Core.getSegmenter());
-        projectTMX.save(props, new File(props.getProjectInternal(), OConsts.TMX_EXTENSION).getPath(), false);
+        projectTMX.save(props, new File(props.getProjectInternal(), OConsts.TMX_EXTENSION).getPath(), false, true);
```
Previous line is even longer and you did not complain...

miurahr replied:
You have a chance to improve code, and I am not complaining but I suggest for you to notify the chance.

miurahr committed:
Merge remote-tracking branch 'origin/master' into fork/t-cordonnier/ProjectTmxLevel2

# Conflicts:
#	src/org/omegat/core/data/ProjectTMX.java
#	src/org/omegat/core/data/RealProject.java
#	src/org/omegat/core/team2/TeamTool.java
#	src/org/omegat/util/TMXReader2.java

miurahr committed:
refactor(core): improve formatting, readability, and logic of TeamTool, ProjectTMX, TMXReader2, and FilterVisitor

- Adjusted line breaks for better code readability.
- Streamlined logic in TMX export by combining conditions.
- Replaced redundant type casts with instance pattern matching.
- Suppressed unchecked warning in XML event parsing.
- Resolved formatting inconsistencies in comments and method definitions.

miurahr committed:
refactor(core): update TMXReader2 imports for explicit XML event references

miurahr committed:
docs(adr): update status to reflect full implementation of TMX validation strategy

miurahr commented on src/org/omegat/core/data/RealProject.java:
```diff
@@ -834,9 +834,10 @@ public synchronized void saveProject(boolean doTeamSync) {
                     }
-                    if (remoteRepositoryProvider.isManaged() && doTeamSync) {
+
+                    if (remoteRepositoryProvider != null && doTeamSync) {
```
`remoteRepositoryProvider` has never been null, so it is always true. It changes the correct `remoteRepositoryProvider.isManaged()` to be degraded.

miurahr commented on src/org/omegat/util/TMXReader2.java:
```diff
@@ -41,11 +41,7 @@
-import javax.xml.stream.events.EndElement;
-import javax.xml.stream.events.StartElement;
-import javax.xml.stream.events.XMLEvent;
+import javax.xml.stream.events.*;
```
it should be individual imports.

miurahr commented on src/org/omegat/core/data/ProjectTMX.java:
```diff
@@ -142,8 +147,8 @@ public boolean isEmpty() {
      */
-    public void save(ProjectProperties props, String translationFile, boolean translationUpdatedByUser)
-            throws Exception {
+    public void save(ProjectProperties props, String translationFile, boolean translationUpdatedByUser,
```
We should have the `save` with the old signature with `@deprecated`

miurahr committed:
refactor(core): enhance imports, formatting, and null safety in TMX and project handling

- Consolidated imports in `RealProject` for readability.
- Improved formatting in loops and comments for consistency.
- Added null checks using `Objects.requireNonNull` for safer code execution.

miurahr committed:
refactor(core): streamline project TMX save logic and improve team project handling

- Introduced overloaded `save` method in `ProjectTMX` for better encapsulation.
- Replaced `remoteRepositoryProvider != null` checks with `remoteRepositoryProvider.isManaged()` because remoteRepositoryProvider always non-null.
- fix team project synchronization conditionals in `RealProject`.

miurahr committed:
Merge branch 'master' into ProjectTmxLevel2

miurahr committed:
refactor(core): organize imports in RealProject for clarity and explicit usage
